### PR TITLE
[multistage] fix decimal types

### PR DESCRIPTION
**Pull request by walterddr**
calcite creates decimal types based on precision of literals. 

- This PR makes it honor precision hoisting.
- Also made error propagation include stack trace.

### CODE REVIEW

File: pinot-query-runtime/src/test/resources/queries/LexicalStructure.json [hunk 1]
```diff
@@ -135,11 +135,18 @@
...
-    "queries": [{
-      "psql": "4.1.2.6",
-      "description": "numeric constant supports",
-      "sql": "SELECT data, 42, 3.5, 4., .001, 5e2, 1.925e-3 FROM {tbl}"
-    }]
+    "queries": [
+      {
+        "psql": "4.1.2.6",
+        "description": "numeric constant supports",
+        "sql": "SELECT data, 42, 3.5, 4., .001, 5e2, 1.925e-3 FROM {tbl}"
```
Comment (agavra) [inline pinot-query-runtime/src/test/resources/queries/LexicalStructure.json:142]: can we add some math functions for these as well? (constant/constant and constant/column)?

**Review by agavra — APPROVED** [review]

File: pinot-query-runtime/src/test/resources/queries/LexicalStructure.json (hunk 1 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/resources/queries/LexicalStructure.json:142]: got it. will add. i wasn't sure if there's anything new to add on top of the one already there. but will do. (it also tests the RelDataTypeFactory's derive type from `+, -, *, /` methods

Comment (walterddr) [reply pinot-query-runtime/src/test/resources/queries/LexicalStructure.json:142]: test added

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-query-runtime/src/test/resources/queries/MathFuncs.json [hunk 2]
```diff
@@ -435,5 +435,95 @@
...
+          {"name": "intCol", "type": "INT"},
+          {"name": "longCol", "type": "LONG"},
+          {"name": "doubleCol", "type": "DOUBLE"},
+          {"name": "floatCol", "type": "FLOAT"}
+        ],
+        "inputs": [
+          [0, 0, 0.1, 3.2],
+          [123, 321, 4.2, 3.0],
+          [-456, -2, 1.1, 7.7],
+          [123, -456, 3.6, 9.1]
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/resources/queries/MathFuncs.json:452]: Curious why we don't have expected results to compare against ?

Comment (walterddr) [reply pinot-query-runtime/src/test/resources/queries/MathFuncs.json:452]: we dont need b/c we are comparing against hard-coded result, b/c we compare against a reference H2 database

**Pull request merged**